### ISSUE TITLE
Can you share pretrained models of CREStereo architectures？thanks

### ISSUE THREAD

**Issue by zhaomeng1028** [bartn8/vppstereo#1]
Can you share pretrained models of CREStereo architectures？thanks

**Comment by bartn8** [comment]
Hi @zhaomeng1028, I downloaded the model from the pytorch implementation of CREStereo.
Link to the model (from README.md): https://drive.google.com/file/d/1D2s1v4VhJlNz98FQpFxf_kBAKQVN_7xo/view
Link to the pytorch repo: https://github.com/ibaiGorordo/CREStereo-Pytorch

**Comment by zhaomeng1028** [comment]
Thanks for your reply，Could you also share fine-tuned models of CREStereo architectures with virtual projected images？

**Comment by bartn8** [comment]
Unfortunately, we did not train CREStereo. We report in Table 6 the experiments done with the aforementioned weights.
If you want to train CREStereo from the PyTorch repo using VPP, you can simply modify the data loader as follows:

CREStereo-Pytorch/dataset.py:1
```python
import os
import cv2
import glob
import numpy as np
from PIL import Image, ImageEnhance
from torch.utils.data import Dataset

#Create sampling function and VPP function
from vpp_standalone import vpp

def sample_hints(hints, probability=0.05):
    new_hints = hints.copy()
    valid_hints = hints>0
    valid_hints = (valid_hints & (np.random.rand(*hints.shape)<probability)).astype(np.float32)
    new_hints[valid_hints==0] = 0
    return new_hints

```

CREStereo-Pytorch/dataset.py:180
```python
    def __getitem__(self, index):
        # find path
        left_path = self.imgs[index]
        prefix = left_path[: left_path.rfind("_")]
        right_path = prefix + "_right.jpg"
        left_disp_path = prefix + "_left.disp.png"
        right_disp_path = prefix + "_right.disp.png"

        # read img, disp
        left_img = cv2.imread(left_path, cv2.IMREAD_COLOR)
        right_img = cv2.imread(right_path, cv2.IMREAD_COLOR)
        left_disp = self.get_disp(left_disp_path)
        right_disp = self.get_disp(right_disp_path)

        if self.rng.binomial(1, 0.5):
            left_img, right_img = np.fliplr(right_img), np.fliplr(left_img)
            left_disp, right_disp = np.fliplr(right_disp), np.fliplr(left_disp)
        left_disp[left_disp == np.inf] = 0

        # augmentaion
        left_img, right_img, left_disp, disp_mask = self.augmentor(
            left_img, right_img, left_disp
        )

        density_steps = np.arange(0, 0.06, 0.01)
        density = np.random.choice(density_steps)
        hints = sample_hints(left_disp,density)

        #Modify VPP parameters as needed
        left_img_vpp, right_img_vpp = vpp(left_img, right_img, hints)

        left_img = left_img.transpose(2, 0, 1).astype("uint8")
        right_img = right_img.transpose(2, 0, 1).astype("uint8")
        left_img_vpp = left_img_vpp.transpose(2, 0, 1).astype("uint8")
        right_img_vpp = right_img_vpp.transpose(2, 0, 1).astype("uint8")

        return {
            "left": left_img_vpp,
            "right": right_img_vpp,
            "left_vanilla": left_img, #Pass also original image
            "right_vanilla": right_img, #Pass also original image
            "disparity": left_disp,
            "mask": disp_mask,
        }
```

I did not try this snippet of code, but the main idea here is to pass virtual projected images during training.

**Comment by asfdgf** [comment]
> Unfortunately, we did not train CREStereo. We report in Table 6 the experiments done with the aforementioned weights. If you want to train CREStereo from the PyTorch repo using VPP, you can simply modify the data loader as follows:
> 
> CREStereo-Pytorch/dataset.py:1
> 
> import os
> import cv2
> import glob
> import numpy as np
> from PIL import Image, ImageEnhance
> from torch.utils.data import Dataset
> 
> #Create sampling function and VPP function
> from vpp_standalone import vpp
> 
> def sample_hints(hints, probability=0.05):
>     new_hints = hints.copy()
>     valid_hints = hints>0
>     valid_hints = (valid_hints & (np.random.rand(*hints.shape)<probability)).astype(np.float32)
>     new_hints[valid_hints==0] = 0
>     return new_hints
> CREStereo-Pytorch/dataset.py:180
> 
>     def __getitem__(self, index):
>         # find path
>         left_path = self.imgs[index]
>         prefix = left_path[: left_path.rfind("_")]
>         right_path = prefix + "_right.jpg"
>         left_disp_path = prefix + "_left.disp.png"
>         right_disp_path = prefix + "_right.disp.png"
> 
>         # read img, disp
>         left_img = cv2.imread(left_path, cv2.IMREAD_COLOR)
>         right_img = cv2.imread(right_path, cv2.IMREAD_COLOR)
>         left_disp = self.get_disp(left_disp_path)
>         right_disp = self.get_disp(right_disp_path)
> 
>         if self.rng.binomial(1, 0.5):
>             left_img, right_img = np.fliplr(right_img), np.fliplr(left_img)
>             left_disp, right_disp = np.fliplr(right_disp), np.fliplr(left_disp)
>         left_disp[left_disp == np.inf] = 0
> 
>         # augmentaion
>         left_img, right_img, left_disp, disp_mask = self.augmentor(
>             left_img, right_img, left_disp
>         )
> 
>         density_steps = np.arange(0, 0.06, 0.01)
>         density = np.random.choice(density_steps)
>         hints = sample_hints(left_disp,density)
> 
>         #Modify VPP parameters as needed
>         left_img_vpp, right_img_vpp = vpp(left_img, right_img, hints)
> 
>         left_img = left_img.transpose(2, 0, 1).astype("uint8")
>         right_img = right_img.transpose(2, 0, 1).astype("uint8")
>         left_img_vpp = left_img_vpp.transpose(2, 0, 1).astype("uint8")
>         right_img_vpp = right_img_vpp.transpose(2, 0, 1).astype("uint8")
> 
>         return {
>             "left": left_img_vpp,
>             "right": right_img_vpp,
>             "left_vanilla": left_img, #Pass also original image
>             "right_vanilla": right_img, #Pass also original image
>             "disparity": left_disp,
>             "mask": disp_mask,
>         }
> I did not try this snippet of code, but the main idea here is to pass virtual projected images during training.

Hey! I am also interested in this project. I'd like to know what exactly is made up of forward input when fine-tuning? Is it incoming Vanilla or VPP enhanced images for image matching?

**Comment by bartn8** [comment]
Hi asfdgf,
I'm glad that you're interested in this project!
During fine-tuning we passed VPP-enhanced images. 
You can look at #3 issue to get more details.

Best,
Luca
